### PR TITLE
Ref count stream state and release when final.

### PR DESCRIPTION
**Pull request by carllerche**
**Work in progress**

Try to actually free memory...

### CODE REVIEW

File: src/proto/streams/stream.rs [hunk 1]
```diff
@@ -1,5 +1,18 @@
...
+///
+/// # Reference counting
+///
+/// There can be a number of outstanding handles to a single Stream. These are
+/// tracked using reference counting. The `ref_count` field represents the
+/// number of outstanding userspace handles that can reach this stream.
+///
+/// It's important to note that when the stream is placed in an internal queue
+/// (such as an accept queue), this is **not** tracked by a reference count.
+/// Thus, `ref_count` can be zero and the stream still has to be kept around.
```
Comment (hawkw) [inline src/proto/streams/stream.rs]: ty for rustdoc :)

**Comment by carllerche** [comment]
Another option besides ref counting would be for `StreamRef` to store both the key (index in the slab) and the stream identifier. Since the stream identifier is monotonically assigned, the `StreamRef` deref fn would lookup the stream state from the slab then ensure that the IDs match.

The downside is that stream reset on drop is harder, but that might be possible to implement statically (by managing type transitions).

**Comment by hawkw** [comment]
@carllerche I, for one, am extremely in favor of representing this w/ type transitions.

File: src/proto/streams/store.rs [hunk 2]
```diff
@@ -170,6 +200,19 @@ impl<B, P> ops::IndexMut<Key> for Store<B, P>
     }
 }
 
+#[cfg(feature = "unstable")]
+impl<B, P> Store<B, P>
```
Comment (olix0r) [inline src/proto/streams/store.rs]: This appears not to work for me. When compiling a project that depends on this:

```
error[E0599]: no method named `num_active_streams` found for type `proto::streams::store::Store<B, P>` in the current scope
   --> /home/ver/.cargo/git/checkouts/h2-5627e19d2c95dbee/4b4aeef/src/proto/streams/streams.rs:440:18
    |
440 |         me.store.num_active_streams()
    |                  ^^^^^^^^^^^^^^^^^^

error[E0599]: no method named `num_wired_streams` found for type `proto::streams::store::Store<B, P>` in the current scope
   --> /home/ver/.cargo/git/checkouts/h2-5627e19d2c95dbee/4b4aeef/src/proto/streams/streams.rs:445:18
    |
445 |         me.store.num_wired_streams()
    |                  ^^^^^^^^^^^^^^^^^
```

File: src/proto/streams/streams.rs [hunk 3]
```diff
@@ -384,6 +430,22 @@ impl<B> Streams<B, client::Peer>
     }
 }
 
+// #[cfg(feature = "unstable")]
```
Comment (olix0r) [inline src/proto/streams/streams.rs]: this needs to be uncommented i think

**Pull request merged**